commands.py

```diff
@@ -248,6 +248,84 @@ def forget(bot,update):
 			catcherror = traceback.format_exc()
 			bot.sendMessage(chat_id=Tokens.channel('errorchannel'),text=catcherror,parse_mode='HTML')
 
+	def list_cattributes(bot,update):
+		try:
+			with closing (pymysql.connect(Tokens.mysql('host'),Tokens.mysql('usn'),Tokens.mysql('pwd'),Tokens.mysql('db'),charset='utf8')) as conn:
+				conn.autocommit(True)
+				with closing(conn.cursor()) as cur:
+					uid = update.message.from_user.id
+					cur.execute("""SELECT * FROM Attribute WHERE telegram_id = %s""",(uid,))
+					if cur.rowcount == 0:
+						message = "You have no cattributes listed!"
+						update.message.reply_text(message,parse_mode='HTML')
+					else:
+						cattributes = cur.fetchall()
+						catlist = [x[1] for x in cattributes]
+						message = "".join(['Your current cattributes are: ',(", ".join(catlist))])
+						update.message.reply_text(message,parse_mode='HTML')
+
+		except Exception as e:
+			catcherror = traceback.format_exc()
+			bot.sendMessage(chat_id=Tokens.channel('errorchannel'),text=catcherror,parse_mode='HTML')
+
+	def remove_cattributes(bot,update,args):
+		try:
+			with closing (pymysql.connect(Tokens.mysql('host'),Tokens.mysql('usn'),Tokens.mysql('pwd'),Tokens.mysql('db'),charset='utf8')) as conn:

```

Comment:
Based on how many times these lines are written, you ought to move the connection to the class instance level, and not open/close it every time.

inside:
```
def __init__(self, tokens):
   self.conn = pymysql.connect(Tokens.connstring)
```

Then you can just keep using the connection, rather than always re-open/close it.
( And call close in the exit handler )

Reply:
Managed to get this to work. However, I can't seem to close the connection - the second time the bot loops through the repeating job, the connection will not open again.